lib/tesla/middleware/telemetry.ex

```diff
@@ -0,0 +1,37 @@
+defmodule Tesla.Middleware.Telemetry do
+  @behaviour Tesla.Middleware
+
+  @moduledoc """
+  Send the request duration and meta-information through telemetry.
+
+  ### Example usage
+  ```
+  defmodule TelemetryHandler do
+    use GenServer
+
+    def start_link(_, _) do
+      :telemetry_app.start(nil, nil)
+      :telemetry.attach(
+        "tesla-telemetry",
+        [:tesla, :telemetry, :traffic],
+        fn ([:tesla, :telemetry, :traffic], time, meta, _config) -> # Do sth end,
+        nil)
+    end
+  end
+
+  defmodule MyClient do
+    use Tesla
+
+    plug Tesla.Middleware.Telemetry
+
+  end
+  ```
+  """
+
+  @doc false
+  def call(env, next, _opts) do
+    {time, res} = :timer.tc(Tesla, :run, [env, next])
+    :telemetry.execute([:tesla, :telemetry, :traffic], time, Enum.into([res], %{}))
```

Comment:
The `Enum.into(...)` part will result in:

```ex
%{ok: %Tesla.Env{status: 200, ...}}
```

or 

```ex
%{error: :nxdomain}
```

I would leave the result as is, like `%{result: {:ok, env}}` or `%{result: {:error, ...}}` to make it consistent (and you can still use nice pattern matching on this structure)